### PR TITLE
[WIP] mxnet_backend graph fix, layer support 

### PR DESCRIPTION
**Pull request by jiajiechen**
===Layers====
* Add Activation Layers 
* Add Pooling Layers

===Fixes===
* Fix name_scoping
* Fix some legacy support such image_dim_ordering
* Removed some hack logic in keras_symbol_child such as reset and REENTRY
* Fix graph generation error

PLEASE DO NOT MERGE BEFORE I ADD THE UNIT TEST. 
@sandeep-krishnamurthy  @yuruofeifei @Roshrini

### CODE REVIEW

File: keras/backend/mxnet_backend.py [hunk 1]
```diff
@@ -167,62 +126,118 @@ def to_dense(tensor):
...
-    """
-    pass
+# class KerasContext(object):
+#     """
+#     TODO: Contexts are not yet supported with MXNet backend.
+#     """
+#     pass
+
+
+def keras_symbol_child(func):
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: Can we rename this to something more intuitive - Ex: keras_to_mxnet_symbol_mapper() or something in similar lines?

File: keras/backend/mxnet_backend.py [hunk 2]
```diff
@@ -167,62 +126,118 @@ def to_dense(tensor):
...
+    """TODO: Add explanation for the keras symbol child."""
+    @wraps(func)
+    def func_wrapper(*args, **kwargs):
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: Looks like test_keras_symbol is not being used at all here.

File: keras/backend/mxnet_backend.py [hunk 3]
```diff
@@ -167,62 +126,118 @@ def to_dense(tensor):
...
-        self._uses_learning_phase = False
-        self._name = name
-        self._neighbor = []
-        for n in neighbor:
-            self.add_neighbor(n)
-        self._bind_values = {}
-        # This will be MXNet NDArray
-        self.tensor = None
+        self._train_sym = mx_symbol if learning_phase() or is_var else None
+        self._pred_sym = mx_symbol if not learning_phase() or is_var else None
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: if not (learning_phase() or is_var) ?

File: keras/backend/mxnet_backend.py [hunk 4]
```diff
@@ -292,7 +309,7 @@ def __abs__(self):
     def __add__(self, other):
         if isinstance(other, KerasSymbol):
             return KerasSymbol(
-                mx.sym.broadcast_add(
+                mx.sym.elemwise_add(
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: Nice job. Thanks for identifying this bug.

File: keras/backend/mxnet_backend.py [hunk 5]
```diff
@@ -472,16 +470,13 @@ def variable(value, dtype=None, name=None, constraint=None):
                [ 3.,  4.]])
     ```
     """
-    if hasattr(value, 'tocoo'):
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: We need this check at the top to make sure sparse tensor is not being created.

File: keras/backend/mxnet_backend.py [hunk 6]
```diff
@@ -1075,8 +1055,7 @@ def random_normal_variable(shape, mean, scale, dtype=None,
     if dtype is None:
         dtype = floatx()
     dtype = _convert_string_dtype(dtype)
-    if name is None:
-        name = _autogen_name("randomnormal")
+    name = _prepare_name(name, "randomnormal")
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: In all these we append to user provided name. Is this expected behavior?

File: keras/backend/mxnet_backend.py [hunk 7]
```diff
@@ -1389,7 +1367,8 @@ def gather(reference, indices):
     # Returns
         A tensor of same type as `reference`.
     """
-    assert ndim(reference) == 2
+    # TODO: this ndim is not used in CNN
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: Please make this to throw NotImplementedError

File: keras/backend/tests/mxnet_backend_test.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+from test_util import *
+from mxnet.base import MXNetError
+
+"""Note: This is a backend test only for using MXNet backend. Many test case are copied from 
+Keras backend test with some modification. The goal is quickly verifying MXNet backend works 
+as expected. We will merge the MXNet specfic test cases into Keras backend tests. 
+"""
+
+
+def tensorflow_clean_session(func):
```
Comment (sandeep-krishnamurthy) [inline keras/backend/tests/mxnet_backend_test.py]: mxnet? :-)

File: pytest.ini [hunk 9]
```diff
@@ -1,7 +1,6 @@
 # Configuration of py.test
 [pytest]
 addopts=-v
-        -n 2
```
Comment (sandeep-krishnamurthy) [inline pytest.ini]: Lets undo this change from commit.

File: keras/backend/mxnet_backend.py [hunk 10]
```diff
@@ -2,97 +2,57 @@
...
-
-            if reset:
-                _REENTRY = False
-            return train_ret
-        finally:
-            if reset:
-                _REENTRY = False
-    return func_wrapper
+@contextmanager
+def name_scope(name):
```
Comment (yuruofeifei) [inline keras/backend/mxnet_backend.py]: This is very cool implementation!

File: keras/backend/mxnet_backend.py [hunk 11]
```diff
@@ -2,97 +2,57 @@
...
+    name += "%d" % get_uid(name)
+    return name
 
 
 def set_model(model):
     global _MODEL
     _MODEL = model
 
 
+def model():
```
Comment (yuruofeifei) [inline keras/backend/mxnet_backend.py]: maybe this can be a python property?

File: keras/backend/mxnet_backend.py [hunk 12]
```diff
@@ -2,97 +2,57 @@
...
+
 def clear_session():
+    global _MODEL
     reset_uids()
-    _EXECUTOR = None
     _MODEL = None
-    _REENTRY = False
 
 
 def learning_phase():
```
Comment (yuruofeifei) [inline keras/backend/mxnet_backend.py]: consider @property ?

File: keras/backend/mxnet_backend.py [hunk 13]
```diff
@@ -167,62 +126,118 @@ def to_dense(tensor):
...
-            if x not in self._neighbor:
-                self._neighbor.append(x)
-                x.add_neighbor(self)
+        if not isinstance(x, KerasSymbol):
+            return
+        if x not in self._neighbors:
+            self._neighbors.append(x)
+            x.add_neighbor(self)
 
     def get_neighbor(self):
```
Comment (yuruofeifei) [inline keras/backend/mxnet_backend.py]: get_neighbors?

File: keras/backend/mxnet_backend.py (hunk 4 above)
Comment (yuruofeifei) [reply keras/backend/mxnet_backend.py]: Why is it not a broadcast add? w * x + b, I think b can be a broadcasted?
